### PR TITLE
[ROCm] Add HIP paged attention kernel for TurboQuant k8v4 decode

### PR DESCRIPTION
**Pull request by andyluo7**
## Summary

Add an optimized HIP paged attention kernel for TurboQuant k8v4 (FP8 key + 4-bit value) decode on AMD ROCm GPUs (MI300X/MI355X).

### What it does

Replaces the Triton-based TQ decode kernel with a HIP kernel on ROCm, using a hybrid dispatch:

- **Per-Q-head kernel** (grid split-K) for low batch (B≤16) or short context
- **MFMA GQA kernel** (`mfma_f32_16x16x16f16`) for high batch + long context — processes all GQA query heads simultaneously, sharing K/V loads

### Files changed (3 files, 805 insertions)

| File | Change |
|------|--------|
| `vllm/v1/attention/ops/tq_k8v4_rocm_decode.hip` | **New** — HIP kernel with v8 split-K + MFMA GQA + stage2 reduction |
| `vllm/v1/attention/ops/tq_k8v4_rocm_decode.py` | **New** — Python wrapper: build/cache .so, dispatch to kernel |
| `vllm/v1/attention/backends/turboquant_attn.py` | **Modified** — ROCm dispatch in `_decode_attention()` |

### What is NOT changed

- Triton TQ decode kernel (untouched)
- CUDA path (guarded by `current_platform.is_rocm()`)
- TQ store/compress ops
- No new environment variables required

### Benchmark Results

**Kernel-level** (MiniMax-M2.5, TP=4, MI355X):
```
 B   ctx   Triton(ms)  HIP(ms)  Speedup
 1   256    0.041      0.015    2.76x
 1  4096    0.045      0.033    1.36x
 8   256    0.041      0.015    2.72x
 8  8192    0.126      0.124    1.02x
32  4096    0.206      0.134    1.54x
64  1024    0.116      0.117    0.99x
```

**E2E serving** (ISL=8192, OSL=1024, TP=4):
```
conc  HIP tok/s  Triton tok/s  Speedup
  4    157.4      131.7        1.20x
  8    284.0      146.2        1.94x
 16    450.2      358.8        1.25x
 32    629.4      383.3        1.64x
 64    728.4      718.0        1.01x
```

### Correctness

| Test | Result |
|------|--------|
| Cosine similarity vs Triton | 0.999996 |
| GSM8K accuracy (15 problems) | 73.3% = Triton 73.3% (identical) |
| MI300X (gfx942) build + correctness | ✅ Pass |
| MI355X (gfx950) full test suite | ✅ Pass |
| NaN/Inf check | ✅ None |

### How it works

1. On ROCm + TQ k8v4 FP8 key: `TurboQuantAttentionImpl.__init__` loads the HIP kernel
2. `_decode_attention()` dispatches to HIP kernel instead of Triton
3. The HIP kernel is compiled at first use via `torch.utils.cpp_extension.load()` and cached
4. Falls back to Triton if HIP kernel build fails

### Test plan

- [x] Kernel correctness: cosine ≥ 0.9999 at B={1,8,32,64}, ctx={256,1024,4096,8192}
- [x] GSM8K accuracy: matches Triton baseline
- [x] MI300X (gfx942): build + correctness verified
- [x] MI355X (gfx950): full benchmark suite
- [x] E2E serving: ISL={1024,8192} sweep
- [x] CUDA path: untouched (ROCm guard)

### CODE REVIEW

File: vllm/v1/attention/ops/tq_k8v4_rocm_decode.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    Returns True if the kernel is loaded and ready.
+    """
+    global _tq_k8v4_loaded
+    if _tq_k8v4_loaded:
+        return True
+
+    if not hasattr(torch.ops, "tq_k8v4") or not hasattr(
+        torch.ops.tq_k8v4, "paged_attention"
+    ):
+        hip_src = Path(__file__).parent / "tq_k8v4_rocm_decode.hip"
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/tq_k8v4_rocm_decode.py]: We should include this into CMakeLists.txt and ship the kernel as prebuilt kernel. Moreover, since the hip is written for gfx942 and gfx950, we have added guards to only use it for gfx942 and gfx950.

**Comment by tjtanaa** [comment]
Please add unit tests to validate the hip kernel against the triton kernel or pytorch reference implementation.

File: vllm/v1/attention/ops/tq_k8v4_rocm_decode.py (hunk 1 above)
Comment (andyluo7) [reply vllm/v1/attention/ops/tq_k8v4_rocm_decode.py]: Done! Added in commit `106389b`:

1. **CMakeLists.txt**: Added `_tq_k8v4_C` extension target with arch-filtered build — iterates `VLLM_GPU_ARCHES` and only compiles for gfx942/gfx950. Kernel source at `csrc/rocm/tq_k8v4_paged_attention.hip`.

2. **Architecture guards**: Python wrapper checks `gcnArchName` and only dispatches to HIP kernel on gfx942/gfx950. Falls back gracefully with a clear warning on other architectures.

3. **Unit tests**: Added `tests/v1/attention/ops/test_tq_k8v4_rocm_decode.py` (~570 lines) with full PyTorch reference implementation, testing multiple batch sizes, sequence lengths, and GQA configs.

**Comment by andyluo7** [comment]
@tjtanaa All addressed in the latest commit:

- ✅ CMakeLists.txt integration with prebuilt kernel (gfx942/gfx950 filtered)
- ✅ Architecture guards in Python wrapper
- ✅ Unit tests (`tests/v1/attention/ops/test_tq_k8v4_rocm_decode.py`, ~570 lines)
- ✅ HEAD_DIM=128 and GQA ratio≤16 validations (Gemini feedback)
- ✅ `torch::empty` → `torch::zeros` for unused-split correctness (Gemini feedback)

**Comment by andyluo7** [comment]
@tjtanaa Addressed in 106389b7 — added CMake build with gfx942/gfx950 arch guards and unit tests validating the HIP kernel against a PyTorch reference implementation. Please take another look when you get a chance!
